Fix #27060: Add indexes identified in constraints to Meta.indexes

drewbrew opened the pull request:
When inspecting tables, the process creates a constraints dict
that identifies indexes. If we have any such indexes, let's add
them to `Meta.indexes` using the base `Index` class.

At this point, it does not support identifying the index type and
substituting in the proper index class, and I don't believe we have
the information at this time to select the right class anyway.

Ref: https://code.djangoproject.com/ticket/27060

drewbrew commented:
Fix for docs test is at https://github.com/django/django/pull/12698.

felixxm commented:
@drewbrew Thanks for this patch :+1: Please take a look at the previous attempt #7083 and all comments in the ticket. This patch should take into account indexes' types and various parameters (e.g. `fillfactor`).

drewbrew commented:
Thanks for that pointer, @felixxm.

So far, I've:

- added in the type check (at least to the level of marking a comment if it's not the default type)
- skipped single-column rtree indexes (and making a note of it in case the assumption that it's a default spatial index is wrong)
- added support for index ordering (if available)

(Tests are still wonky, I know)

I'm a little confused by the postgres-specific options. My initial thought is that I'm going to have to do some sort of mapping from `Index` subclassses' `type` to that class (in the postgres introspection module) and convert the the options as returned from the db into kwargs accordingly.  Does that sound like a reasonable and maintainable approach?

drewbrew commented:
Thanks for the prod, @nessita (and congrats on being the new fellow!). I've rebased it (correctly I hope... a lot changes in 3 years!), but how do I flag for review? Just convert it to a draft and back?

nessita commented:
Thank you @drewbrew. Could you please revert the changes you made to AUTHORS (other than adding your name to it)?
